File: data_mcp/gpt-5/go/rep1/main.go
package main

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"log"
	"net/http"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Data models

type User struct {
	ID       int    `json:"id"`
	Username string `json:"username"`
	Password string `json:"-"`
}

type Todo struct {
	ID          int    `json:"id"`
	Title       string `json:"title"`
	Description string `json:"description"`
	Completed   bool   `json:"completed"`
	CreatedAt   string `json:"created_at"`
	UpdatedAt   string `json:"updated_at"`
	UserID      int    `json:"-"`
}

type Store struct {
	mu          sync.RWMutex
	usersByID   map[int]*User
	usersByName map[string]*User
	sessions    map[string]int // token -> userID
	todosByID   map[int]*Todo
	nextUserID  int
	nextTodoID  int
}

func NewStore() *Store {
	return &Store{
		usersByID:   make(map[int]*User),
		usersByName: make(map[string]*User),
		sessions:    make(map[string]int),
		todosByID:   make(map[int]*Todo),
		nextUserID:  1,
		nextTodoID:  1,
	}
}

// Helpers

func writeJSON(w http.ResponseWriter, status int, v interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	enc := json.NewEncoder(w)
	enc.SetEscapeHTML(true)
	_ = enc.Encode(v)
}

func writeError(w http.ResponseWriter, status int, msg string) {
	writeJSON(w, status, map[string]string{"error": msg})
}

func readJSON(r *http.Request, dst interface{}) error {
	dec := json.NewDecoder(r.Body)
	dec.DisallowUnknownFields()
	return dec.Decode(dst)
}

func generateToken() (string, error) {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return hex.EncodeToString(b), nil
}

func nowISO8601() string {
	return time.Now().UTC().Format(time.RFC3339)
}

// Authentication

func (s *Store) getUserFromRequest(r *http.Request) (*User, string, error) {
	c, err := r.Cookie("session_id")
	if err != nil || c.Value == "" {
		return nil, "", errors.New("auth required")
	}
	s.mu.RLock()
	uid, ok := s.sessions[c.Value]
	if !ok {
		s.mu.RUnlock()
		return nil, c.Value, errors.New("auth required")
	}
	user := s.usersByID[uid]
	s.mu.RUnlock()
	if user == nil {
		return nil, c.Value, errors.New("auth required")
	}
	return user, c.Value, nil
}

// Handlers

type Server struct {
	store *Store
}

func NewServer() *Server {
	return &Server{store: NewStore()}
}

var usernameRe = regexp.MustCompile(`^[a-zA-Z0-9_]{3,50}$`)

func (s *Server) handleRegister(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	type reqT struct {
		Username string `json:"username"`
		Password string `json:"password"`
	}
	var req reqT
	if err := readJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}
	if !usernameRe.MatchString(req.Username) {
		writeError(w, http.StatusBadRequest, "Invalid username")
		return
	}
	if len(req.Password) < 8 {
		writeError(w, http.StatusBadRequest, "Password too short")
		return
	}
	s.store.mu.Lock()
	defer s.store.mu.Unlock()
	if _, exists := s.store.usersByName[req.Username]; exists {
		writeError(w, http.StatusConflict, "Username already exists")
		return
	}
	id := s.store.nextUserID
	s.store.nextUserID++
	user := &User{ID: id, Username: req.Username, Password: req.Password}
	s.store.usersByID[id] = user
	s.store.usersByName[req.Username] = user
	writeJSON(w, http.StatusCreated, map[string]interface{}{"id": user.ID, "username": user.Username})
}

func (s *Server) handleLogin(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	type reqT struct {
		Username string `json:"username"`
		Password string `json:"password"`
	}
	var req reqT
	if err := readJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}
	s.store.mu.RLock()
	user := s.store.usersByName[req.Username]
	s.store.mu.RUnlock()
	if user == nil || user.Password != req.Password {
		writeError(w, http.StatusUnauthorized, "Invalid credentials")
		return
	}
	tok, err := generateToken()
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Server error")
		return
	}
	s.store.mu.Lock()
	s.store.sessions[tok] = user.ID
	s.store.mu.Unlock()
	http.SetCookie(w, &http.Cookie{Name: "session_id", Value: tok, Path: "/", HttpOnly: true})
	writeJSON(w, http.StatusOK, map[string]interface{}{"id": user.ID, "username": user.Username})
}

func (s *Server) handleLogout(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	user, token, err := s.store.getUserFromRequest(r)
	_ = user
	if err != nil {
		writeError(w, http.StatusUnauthorized, "Authentication required")
		return
	}
	s.store.mu.Lock()
	delete(s.store.sessions, token)
	s.store.mu.Unlock()
	// Expire the cookie client-side as well
	http.SetCookie(w, &http.Cookie{Name: "session_id", Value: "", Path: "/", HttpOnly: true, MaxAge: -1, Expires: time.Unix(0, 0)})
	writeJSON(w, http.StatusOK, map[string]interface{}{})
}

func (s *Server) handleMe(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	user, _, err := s.store.getUserFromRequest(r)
	if err != nil {
		writeError(w, http.StatusUnauthorized, "Authentication required")
		return
	}
	writeJSON(w, http.StatusOK, map[string]interface{}{"id": user.ID, "username": user.Username})
}

func (s *Server) handlePassword(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPut {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	user, _, err := s.store.getUserFromRequest(r)
	if err != nil {
		writeError(w, http.StatusUnauthorized, "Authentication required")
		return
	}
	type reqT struct {
		OldPassword string `json:"old_password"`
		NewPassword string `json:"new_password"`
	}
	var req reqT
	if err := readJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}
	if req.OldPassword != user.Password {
		writeError(w, http.StatusUnauthorized, "Invalid credentials")
		return
	}
	if len(req.NewPassword) < 8 {
		writeError(w, http.StatusBadRequest, "Password too short")
		return
	}
	s.store.mu.Lock()
	user.Password = req.NewPassword
	s.store.mu.Unlock()
	writeJSON(w, http.StatusOK, map[string]interface{}{})
}

func (s *Server) handleTodos(w http.ResponseWriter, r *http.Request) {
	user, _, err := s.store.getUserFromRequest(r)
	if err != nil {
		writeError(w, http.StatusUnauthorized, "Authentication required")
		return
	}
	switch r.Method {
	case http.MethodGet:
		// List todos for user ordered by id asc
		s.store.mu.RLock()
		var list []*Todo
		for _, t := range s.store.todosByID {
			if t.UserID == user.ID {
				// make a copy to avoid accidental mutation
				copyT := *t
				list = append(list, &copyT)
			}
		}
		s.store.mu.RUnlock()
		sort.Slice(list, func(i, j int) bool { return list[i].ID < list[j].ID })
		writeJSON(w, http.StatusOK, list)
	case http.MethodPost:
		type reqT struct {
			Title       string `json:"title"`
			Description string `json:"description"`
		}
		var req reqT
		if err := readJSON(r, &req); err != nil {
			writeError(w, http.StatusBadRequest, "Invalid JSON")
			return
		}
		if strings.TrimSpace(req.Title) == "" {
			writeError(w, http.StatusBadRequest, "Title is required")
			return
		}
		if req.Description == "" {
			// default already empty
		}
		created := nowISO8601()
		s.store.mu.Lock()
		id := s.store.nextTodoID
		s.store.nextTodoID++
		todo := &Todo{
			ID:          id,
			Title:       req.Title,
			Description: req.Description,
			Completed:   false,
			CreatedAt:   created,
			UpdatedAt:   created,
			UserID:      user.ID,
		}
		s.store.todosByID[id] = todo
		s.store.mu.Unlock()
		writeJSON(w, http.StatusCreated, todo)
	default:
		writeError(w, http.StatusNotFound, "Not found")
	}
}

func (s *Server) handleTodoItem(w http.ResponseWriter, r *http.Request) {
	user, _, err := s.store.getUserFromRequest(r)
	if err != nil {
		// For any method, require auth
		writeError(w, http.StatusUnauthorized, "Authentication required")
		return
	}
	// Path expected: /todos/{id}
	parts := strings.Split(strings.TrimPrefix(r.URL.Path, "/todos/"), "/")
	if len(parts) == 0 || parts[0] == "" || (len(parts) > 1 && parts[1] != "") {
		// malformed or extra segments
		writeError(w, http.StatusNotFound, "Todo not found")
		return
	}
	idStr := parts[0]
	id, err := strconv.Atoi(idStr)
	if err != nil || id <= 0 {
		writeError(w, http.StatusNotFound, "Todo not found")
		return
	}

	s.store.mu.RLock()
	t, ok := s.store.todosByID[id]
	if ok {
		// ensure ownership
		if t.UserID != user.ID {
			// do not reveal existence
			s.store.mu.RUnlock()
			if r.Method == http.MethodDelete {
				// 404 with JSON per spec, but DELETE success returns no body; here not success
				writeError(w, http.StatusNotFound, "Todo not found")
				return
			}
			writeError(w, http.StatusNotFound, "Todo not found")
			return
		}
	}
	s.store.mu.RUnlock()
	if !ok {
		writeError(w, http.StatusNotFound, "Todo not found")
		return
	}

	switch r.Method {
	case http.MethodGet:
		// return copy without UserID
		copyT := *t
		writeJSON(w, http.StatusOK, &copyT)
	case http.MethodPut:
		type reqT struct {
			Title       *string `json:"title"`
			Description *string `json:"description"`
			Completed   *bool   `json:"completed"`
		}
		var req reqT
		if err := readJSON(r, &req); err != nil {
			writeError(w, http.StatusBadRequest, "Invalid JSON")
			return
		}
		// Validate title if present
		if req.Title != nil && strings.TrimSpace(*req.Title) == "" {
			writeError(w, http.StatusBadRequest, "Title is required")
			return
		}
		s.store.mu.Lock()
		// Re-check to avoid race
		cur, ok2 := s.store.todosByID[id]
		if !ok2 || cur.UserID != user.ID {
			s.store.mu.Unlock()
			writeError(w, http.StatusNotFound, "Todo not found")
			return
		}
		if req.Title != nil {
			cur.Title = *req.Title
		}
		if req.Description != nil {
			cur.Description = *req.Description
		}
		if req.Completed != nil {
			cur.Completed = *req.Completed
		}
		cur.UpdatedAt = nowISO8601()
		updated := *cur
		s.store.mu.Unlock()
		writeJSON(w, http.StatusOK, &updated)
	case http.MethodDelete:
		// Delete and return 204 with no body
		s.store.mu.Lock()
		// Re-check ownership
		cur, ok2 := s.store.todosByID[id]
		if !ok2 || cur.UserID != user.ID {
			s.store.mu.Unlock()
			writeError(w, http.StatusNotFound, "Todo not found")
			return
		}
		delete(s.store.todosByID, id)
		s.store.mu.Unlock()
		w.WriteHeader(http.StatusNoContent)
		return
	default:
		writeError(w, http.StatusNotFound, "Todo not found")
	}
}

func (s *Server) routes() http.Handler {
	mux := http.NewServeMux()
	mux.HandleFunc("/register", s.handleRegister)
	mux.HandleFunc("/login", s.handleLogin)
	mux.HandleFunc("/logout", s.handleLogout)
	mux.HandleFunc("/me", s.handleMe)
	mux.HandleFunc("/password", s.handlePassword)
	mux.HandleFunc("/todos", s.handleTodos)   // GET, POST
	mux.HandleFunc("/todos/", s.handleTodoItem) // GET, PUT, DELETE
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Ensure JSON Content-Type for all responses except DELETE success where handler returns without writing
		// We cannot set header here globally because handlers may return 204 without body.
		// We only ensure unknown routes return JSON error.
		// Delegate to mux and intercept 404s by checking if a handler exists.
		rec := &responseRecorder{ResponseWriter: w}
		mux.ServeHTTP(rec, r)
		if rec.notFound {
			writeError(w, http.StatusNotFound, "Not found")
		}
	})
}

type responseRecorder struct {
	http.ResponseWriter
	notFound bool
}

func (rr *responseRecorder) WriteHeader(statusCode int) {
	if statusCode == http.StatusNotFound {
		rr.notFound = true
		// do not write yet; caller will handle
		return
	}
	rr.ResponseWriter.WriteHeader(statusCode)
}

func (rr *responseRecorder) Write(b []byte) (int, error) {
	if rr.notFound {
		// swallow
		return len(b), nil
	}
	return rr.ResponseWriter.Write(b)
}

func main() {
	var port string
	flag.StringVar(&port, "port", "8080", "port to listen on")
	// allow --port also by default flag pkg supports it
	flag.Parse()
	addr := fmt.Sprintf("0.0.0.0:%s", port)

	server := NewServer()
	mux := server.routes()

	httpServer := &http.Server{
		Addr:              addr,
		Handler:           mux,
		ReadHeaderTimeout: 5 * time.Second,
	}
	log.Printf("Listening on %s", addr)
	if err := httpServer.ListenAndServe(); err != nil && err != http.ErrServerClosed {
		log.Fatalf("server error: %v", err)
	}
}

File: data_mcp/gpt-5/java/rep1/src/Main.java
import com.sun.net.httpserver.*;
import java.io.*;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.time.*;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.regex.*;
import com.google.gson.*;

public class Main {
    // Data models
    static class User {
        final int id;
        final String username;
        String password; // stored in-memory plain for this exercise
        User(int id, String username, String password) {
            this.id = id;
            this.username = username;
            this.password = password;
        }
        Map<String,Object> toPublicJson() {
            Map<String,Object> m = new LinkedHashMap<>();
            m.put("id", id);
            m.put("username", username);
            return m;
        }
    }

    static class Todo {
        final int id;
        final int userId;
        String title;
        String description;
        boolean completed;
        String created_at;
        String updated_at;
        Todo(int id, int userId, String title, String description, boolean completed, String created_at, String updated_at) {
            this.id = id;
            this.userId = userId;
            this.title = title;
            this.description = description;
            this.completed = completed;
            this.created_at = created_at;
            this.updated_at = updated_at;
        }
        Map<String,Object> toJson() {
            Map<String,Object> m = new LinkedHashMap<>();
            m.put("id", id);
            m.put("title", title);
            m.put("description", description);
            m.put("completed", completed);
            m.put("created_at", created_at);
            m.put("updated_at", updated_at);
            return m;
        }
    }

    static class DB {
        final ConcurrentHashMap<String, User> usersByUsername = new ConcurrentHashMap<>();
        final ConcurrentHashMap<Integer, User> usersById = new ConcurrentHashMap<>();
        final AtomicInteger userIdSeq = new AtomicInteger(1);

        final ConcurrentHashMap<Integer, Todo> todosById = new ConcurrentHashMap<>();
        final AtomicInteger todoIdSeq = new AtomicInteger(1);

        final ConcurrentHashMap<String, Integer> sessions = new ConcurrentHashMap<>(); // token -> userId
    }

    static final DB db = new DB();
    static final Gson gson = new GsonBuilder().serializeNulls().create();
    static final Pattern USERNAME_PATTERN = Pattern.compile("^[a-zA-Z0-9_]{3,50}$");

    public static void main(String[] args) throws Exception {
        int port = 8080;
        for (int i=0; i<args.length; i++) {
            if ("--port".equals(args[i]) && i+1 < args.length) {
                port = Integer.parseInt(args[i+1]);
                i++;
            }
        }
        InetSocketAddress addr = new InetSocketAddress("0.0.0.0", port);
        HttpServer server = HttpServer.create(addr, 0);
        HttpContext root = server.createContext("/", Main::handleRoot);
        // Use a thread pool
        server.setExecutor(Executors.newCachedThreadPool());
        server.start();
        System.out.println("Server started on 0.0.0.0:" + port);
    }

    static void handleRoot(HttpExchange ex) throws IOException {
        try {
            String method = ex.getRequestMethod();
            URI uri = ex.getRequestURI();
            String path = uri.getPath();
            if (path.equals("/register") && method.equals("POST")) {
                handleRegister(ex);
                return;
            }
            if (path.equals("/login") && method.equals("POST")) {
                handleLogin(ex);
                return;
            }
            if (path.equals("/logout") && method.equals("POST")) {
                withAuth(ex, (user, token) -> handleLogout(ex, user, token));
                return;
            }
            if (path.equals("/me") && method.equals("GET")) {
                withAuth(ex, (user, token) -> sendJson(ex, 200, user.toPublicJson()));
                return;
            }
            if (path.equals("/password") && method.equals("PUT")) {
                withAuth(ex, (user, token) -> handlePasswordChange(ex, user));
                return;
            }
            if (path.startsWith("/todos")) {
                withAuth(ex, (user, token) -> handleTodos(ex, user));
                return;
            }
            sendError(ex, 404, "Not found");
        } catch (Exception e) {
            e.printStackTrace();
            sendError(ex, 500, "Internal server error");
        }
    }

    @FunctionalInterface
    interface AuthedHandler { void handle(User user, String sessionToken) throws IOException; }

    static void withAuth(HttpExchange ex, AuthedHandler handler) throws IOException {
        String token = getSessionToken(ex.getRequestHeaders());
        if (token == null) {
            sendAuthRequired(ex);
            return;
        }
        Integer uid = db.sessions.get(token);
        if (uid == null) {
            sendAuthRequired(ex);
            return;
        }
        User user = db.usersById.get(uid);
        if (user == null) {
            // session points to missing user; treat as unauthenticated
            sendAuthRequired(ex);
            return;
        }
        handler.handle(user, token);
    }

    static void handleRegister(HttpExchange ex) throws IOException {
        String body = readBody(ex);
        JsonObject req;
        try {
            req = gson.fromJson(body, JsonObject.class);
        } catch (JsonSyntaxException jse) {
            sendError(ex, 400, "Invalid JSON");
            return;
        }
        if (req == null) {
            sendError(ex, 400, "Invalid JSON");
            return;
        }
        String username = getAsString(req.get("username"));
        String password = getAsString(req.get("password"));
        if (username == null || !USERNAME_PATTERN.matcher(username).matches()) {
            sendError(ex, 400, "Invalid username");
            return;
        }
        if (password == null || password.length() < 8) {
            sendError(ex, 400, "Password too short");
            return;
        }
        // Uniqueness
        synchronized (db) {
            if (db.usersByUsername.containsKey(username)) {
                sendError(ex, 409, "Username already exists");
                return;
            }
            int id = db.userIdSeq.getAndIncrement();
            User user = new User(id, username, password);
            db.usersByUsername.put(username, user);
            db.usersById.put(id, user);
            sendJson(ex, 201, user.toPublicJson());
        }
    }

    static void handleLogin(HttpExchange ex) throws IOException {
        String body = readBody(ex);
        JsonObject req;
        try {
            req = gson.fromJson(body, JsonObject.class);
        } catch (JsonSyntaxException jse) {
            sendError(ex, 400, "Invalid JSON");
            return;
        }
        if (req == null) {
            sendError(ex, 400, "Invalid JSON");
            return;
        }
        String username = getAsString(req.get("username"));
        String password = getAsString(req.get("password"));
        User user = username != null ? db.usersByUsername.get(username) : null;
        if (user == null || password == null || !safeEquals(user.password, password)) {
            sendError(ex, 401, "Invalid credentials");
            return;
        }
        String token = UUID.randomUUID().toString().replace("-", "");
        db.sessions.put(token, user.id);
        // Set-Cookie header
        ex.getResponseHeaders().add("Set-Cookie", "session_id=" + token + "; Path=/; HttpOnly");
        sendJson(ex, 200, user.toPublicJson());
    }

    static void handleLogout(HttpExchange ex, User user, String token) throws IOException {
        if (token != null) {
            db.sessions.remove(token);
        }
        sendJson(ex, 200, new LinkedHashMap<String,Object>()); // {}
    }

    static void handlePasswordChange(HttpExchange ex, User user) throws IOException {
        String body = readBody(ex);
        JsonObject req;
        try {
            req = gson.fromJson(body, JsonObject.class);
        } catch (JsonSyntaxException jse) {
            sendError(ex, 400, "Invalid JSON");
            return;
        }
        if (req == null) {
            sendError(ex, 400, "Invalid JSON");
            return;
        }
        String oldp = getAsString(req.get("old_password"));
        String newp = getAsString(req.get("new_password"));
        if (oldp == null || !safeEquals(oldp, user.password)) {
            sendError(ex, 401, "Invalid credentials");
            return;
        }
        if (newp == null || newp.length() < 8) {
            sendError(ex, 400, "Password too short");
            return;
        }
        user.password = newp;
        sendJson(ex, 200, new LinkedHashMap<String,Object>()); // {}
    }

    static void handleTodos(HttpExchange ex, User user) throws IOException {
        String method = ex.getRequestMethod();
        String path = ex.getRequestURI().getPath();
        if (path.equals("/todos")) {
            if (method.equals("GET")) {
                // list
                List<Todo> list = new ArrayList<>();
                for (Todo t : db.todosById.values()) {
                    if (t.userId == user.id) list.add(t);
                }
                list.sort(Comparator.comparingInt(t -> t.id));
                List<Map<String,Object>> out = new ArrayList<>();
                for (Todo t : list) out.add(t.toJson());
                sendJson(ex, 200, out);
                return;
            } else if (method.equals("POST")) {
                String body = readBody(ex);
                JsonObject req;
                try { req = gson.fromJson(body, JsonObject.class); } catch (JsonSyntaxException jse) { sendError(ex, 400, "Invalid JSON"); return; }
                if (req == null) { sendError(ex, 400, "Invalid JSON"); return; }
                String title = getAsString(req.get("title"));
                if (title == null || title.trim().isEmpty()) {
                    sendError(ex, 400, "Title is required");
                    return;
                }
                String description = req.has("description") && !req.get("description").isJsonNull() ? getAsString(req.get("description")) : "";
                if (description == null) description = "";
                String now = isoNow();
                int id = db.todoIdSeq.getAndIncrement();
                Todo t = new Todo(id, user.id, title, description, false, now, now);
                db.todosById.put(id, t);
                sendJson(ex, 201, t.toJson());
                return;
            } else {
                sendError(ex, 405, "Method not allowed");
                return;
            }
        } else if (path.startsWith("/todos/")) {
            String rest = path.substring("/todos/".length());
            Integer tid = parseIntOrNull(rest);
            if (tid == null) { sendError(ex, 404, "Todo not found"); return; }
            Todo t = db.todosById.get(tid);
            if (t == null || t.userId != user.id) {
                sendError(ex, 404, "Todo not found");
                return;
            }
            switch (method) {
                case "GET":
                    sendJson(ex, 200, t.toJson());
                    return;
                case "PUT": {
                    String body = readBody(ex);
                    JsonObject req;
                    try { req = gson.fromJson(body, JsonObject.class); } catch (JsonSyntaxException jse) { sendError(ex, 400, "Invalid JSON"); return; }
                    if (req == null) { sendError(ex, 400, "Invalid JSON"); return; }
                    if (req.has("title") && !req.get("title").isJsonNull()) {
                        String title = getAsString(req.get("title"));
                        if (title == null || title.trim().isEmpty()) { sendError(ex, 400, "Title is required"); return; }
                        t.title = title;
                    } else if (req.has("title") && req.get("title").isJsonNull()) {
                        // title explicitly null is treated as invalid (empty)
                        sendError(ex, 400, "Title is required");
                        return;
                    }
                    if (req.has("description")) {
                        if (req.get("description").isJsonNull()) {
                            t.description = "";
                        } else {
                            String d = getAsString(req.get("description"));
                            t.description = (d == null) ? t.description : d;
                        }
                    }
                    if (req.has("completed")) {
                        JsonElement ce = req.get("completed");
                        if (ce.isJsonNull()) {
                            // leave unchanged
                        } else if (ce.isJsonPrimitive() && ((JsonPrimitive)ce).isBoolean()) {
                            t.completed = ce.getAsBoolean();
                        } else {
                            // ignore invalid type
                        }
                    }
                    t.updated_at = isoNow();
                    sendJson(ex, 200, t.toJson());
                    return;
                }
                case "DELETE":
                    db.todosById.remove(t.id);
                    sendNoContent(ex);
                    return;
                default:
                    sendError(ex, 405, "Method not allowed");
                    return;
            }
        } else {
            sendError(ex, 404, "Not found");
        }
    }

    static String getSessionToken(Headers headers) {
        List<String> cookies = headers.get("Cookie");
        if (cookies == null) return null;
        for (String headerVal : cookies) {
            String[] parts = headerVal.split(";\\s*");
            for (String p : parts) {
                int eq = p.indexOf('=');
                if (eq > 0) {
                    String name = p.substring(0, eq).trim();
                    String val = p.substring(eq+1).trim();
                    if (name.equals("session_id")) return val;
                }
            }
        }
        return null;
    }

    static void sendJson(HttpExchange ex, int status, Object obj) throws IOException {
        byte[] data = gson.toJson(obj).getBytes(StandardCharsets.UTF_8);
        Headers h = ex.getResponseHeaders();
        h.set("Content-Type", "application/json");
        ex.sendResponseHeaders(status, data.length);
        try (OutputStream os = ex.getResponseBody()) { os.write(data); }
    }

    static void sendNoContent(HttpExchange ex) throws IOException {
        // 204 with no body and no content-type header as per spec exception
        ex.sendResponseHeaders(204, -1);
        ex.close();
    }

    static void sendError(HttpExchange ex, int status, String message) throws IOException {
        Map<String, String> m = new LinkedHashMap<>();
        m.put("error", message);
        sendJson(ex, status, m);
    }

    static void sendAuthRequired(HttpExchange ex) throws IOException {
        sendError(ex, 401, "Authentication required");
    }

    static String readBody(HttpExchange ex) throws IOException {
        try (InputStream is = ex.getRequestBody()) {
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            byte[] buf = new byte[4096];
            int r;
            while ((r = is.read(buf)) != -1) baos.write(buf, 0, r);
            return baos.toString(StandardCharsets.UTF_8);
        }
    }

    static String getAsString(JsonElement e) {
        if (e == null || e.isJsonNull()) return null;
        if (e.isJsonPrimitive()) {
            JsonPrimitive p = e.getAsJsonPrimitive();
            if (p.isString()) return p.getAsString();
            // if number or boolean provided, coerce to string
            return p.getAsString();
        }
        return null;
    }

    static Integer parseIntOrNull(String s) {
        try { return Integer.parseInt(s); } catch (Exception e) { return null; }
    }

    static String isoNow() {
        return Instant.now().truncatedTo(ChronoUnit.SECONDS).toString(); // ISO-8601, ends with 'Z'
    }

    static boolean safeEquals(String a, String b) {
        if (a == null || b == null) return false;
        if (a.length() != b.length()) return false;
        int r = 0;
        for (int i=0; i<a.length(); i++) r |= a.charAt(i) ^ b.charAt(i);
        return r == 0;
    }
}
